action/action-util/src/main/java/com/powsybl/action/util/ProportionalScalable.java

```diff
@@ -25,30 +99,65 @@ private static void checkPercentages(List<Float> percentages, List<Scalable> sca
+        return scalablePercentageList.stream().anyMatch(ScalablePercentage::notSaturated);
+    }
 
-    ProportionalScalable(List<Float> percentages, List<Scalable> scalables) {
-        super(scalables);
-        checkPercentages(percentages, scalables);
-        this.percentages = Objects.requireNonNull(percentages);
+    private void checkIterationPercentages() {
+        double iterationPercentagesSum = scalablePercentageList.stream().mapToDouble(ScalablePercentage::getIterationPercentage).sum();
+        if (Math.abs(100 - iterationPercentagesSum) > 0.01) {
+            throw new AssertionError(String.format("Error in proportional scalable ventilation. Sum of percentages must be equals to 100 (%.2f)", iterationPercentagesSum));
+        }
+    }
+
+    private void updateIterationPercentages() {
+        double unsaturatedPercentagesSum = scalablePercentageList.stream().filter(ScalablePercentage::notSaturated).mapToDouble(ScalablePercentage::getIterationPercentage).sum();
+        scalablePercentageList.forEach(scalablePercentage -> {
+            if (scalablePercentage.isSaturated()) {
+                scalablePercentage.setIterationPercentage(0);
+            } else {
+                scalablePercentage.setIterationPercentage(scalablePercentage.getPercentage() / unsaturatedPercentagesSum * 100);
+            }
+        });
+    }
+
+    private double iterativeScale(Network n, double asked, ScalingConvention scalingConvention) {
+        double done = 0;
+        while (Math.abs(asked - done) > 1e-5 && notSaturated()) {
+            checkIterationPercentages();
+            done += atomicScale(n, asked - done, scalingConvention);
+            updateIterationPercentages();
+        }
+        return done;
+    }
+
+    private double atomicScale(Network n, double asked, ScalingConvention scalingConvention) {
+        AtomicDouble done = new AtomicDouble(0);
+        scalablePercentageList.forEach(scalablePercentage -> {
+            Scalable s = scalablePercentage.getScalable();
+            double p = scalablePercentage.getIterationPercentage();
+            double doneOnScalable = s.scale(n, p / 100 * asked, scalingConvention);
+            if (Math.abs(doneOnScalable - p) > 1e-2) {
+                scalablePercentage.setSaturated(true);
+            }
+            done.getAndAdd(doneOnScalable);
+        });
+        return done.get();
     }
 
     @Override
     public double scale(Network n, double asked, ScalingConvention scalingConvention) {
         Objects.requireNonNull(n);
         Objects.requireNonNull(scalingConvention);
-
-        double done = 0;
-        for (int i = 0; i < scalables.size(); i++) {
-            Scalable s = scalables.get(i);
-            double p = percentages.get(i);
-            done += s.scale(n, p / 100 * asked, scalingConvention);
+        if (iterative) {
+            return iterativeScale(n, asked, scalingConvention);
+        } else {
+            return atomicScale(n, asked, scalingConvention);
```

Comment:
I am not sure atomic is the right term. oneIterationScale? simpleScale?

Reply:
What do you think of recursiveScale and scaleIteration ?

Reply:
renamed